### PR TITLE
[quant] Implement PTQ for APoT FakeQuant

### PR DESCRIPTION
**Pull request by asl3**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #81040

### Summary:
This PR implements PTQ for APoT FakeQuant. It runs models (Resnet-18 pre-trained model, ImageNet dataset) to compare accuracy metrics for different qconfig settings of uniform vs. APoT quantized activation and weight. 

According to the collected accuracy stats, model #2 (uniform activation and APoT weight) appears to have a slight improvement in accuracy compared to model #1 (uniform activation and uniform weight) for 8-bit and significant improvement for 4-bit (see "Accuracy Stats" section below).

### Test Plan:
Run models with: `python test/quantization/core/experimental/fx_graph_mode_apot.py`

### Accuracy Stats: 
8-bit (Uniform int8, APoT b = 8 k = 2)

**Model #1:** Uniform activation, uniform weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.43% (Top-1), 85.62% (Top-5)

**Model #2:** Uniform activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.51% (Top-1), 85.78% (Top-5)

**Model #3:** APoT activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.32% (Top-1), 85.78% (Top-5)

4-bit (Uniform int4, APoT b = 4 k = 2)

**Model #1:** Uniform activation, uniform weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 45.63% (Top-1), 71.96% (Top-5)

**Model #2:** Uniform activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 64.24% (Top-1), 85.56% (Top-5)

**Model #3:** APoT activation, APoT weight (FX Graph Mode quantized)
Evaluation accuracy on test dataset: 45.40% (Top-1), 76.21% (Top-5)

**Full Precision model (FX Graph Mode quantized)**
Evaluation accuracy on test dataset: 69.76% (Top-1), 89.08% (Top-5)

**Eager mode quantized model**
Evaluation accuracy on test dataset: 69.49% (Top-1), 88.90% (Top-5)

### CODE REVIEW

**Comment by dzdang** [comment]
this PR is for PTQ and not QAT?

File: test/quantization/core/experimental/fx_graph_mode_apot.py [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+"""
+
+qconfig1 = get_uniform_qconfig_4bit()
+qconfig_dict1 = {"object_type": [(torch.nn.Linear, qconfig1)]}
+prepared_model1 = prepare_fx(float_model, qconfig_dict1)  # fuse modules and insert observers
+calibrate(prepared_model1, data_loader_test)  # run calibration on sample data
+quantized_model1 = convert_fx(prepared_model1)  # convert the calibrated model to a quantized model
+
+top1, top5 = evaluate(quantized_model1, criterion, data_loader_test)
+print("Model #1 Evaluation accuracy on test dataset (b=4, k=2): %2.2f, %2.2f" % (top1.avg, top5.avg))
```
Comment (jerryzh168) [inline test/quantization/core/experimental/fx_graph_mode_apot.py:204]: can we define a helper function for these things?

File: test/quantization/core/experimental/fx_graph_mode_apot.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+print("Model #1 Evaluation accuracy on test dataset (b=8, k=2): %2.2f, %2.2f" % (top1.avg, top5.avg))
+
+"""
+Prepare model with uniform activation, uniform weight
+b=4, k=2
+"""
+
+qconfig1 = get_uniform_qconfig_4bit()
+qconfig_dict1 = {"object_type": [(torch.nn.Linear, qconfig1)]}
+prepared_model1 = prepare_fx(float_model, qconfig_dict1)  # fuse modules and insert observers
```
Comment (jerryzh168) [inline test/quantization/core/experimental/fx_graph_mode_apot.py:199]: do we need to do a deepcopy for float_model?

File: test/quantization/core/experimental/fx_graph_mode_apot.py [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+quantized_model3 = convert_fx(prepared_model3)  # convert the calibrated model to a quantized model
+
+top1, top5 = evaluate(quantized_model3, criterion, data_loader_test)
+print("Model #3 Evaluation accuracy on test dataset (b=4, k=2): %2.2f, %2.2f" % (top1.avg, top5.avg))
+
+"""
+Prepare eager mode quantized model
+"""
+
+from torchvision.models.quantization.resnet import resnet18
```
Comment (jerryzh168) [inline test/quantization/core/experimental/fx_graph_mode_apot.py:267]: would be good to put this to a separate function as well

File: torch/ao/quantization/experimental/apot_utils.py [hunk 4]
```diff
@@ -5,52 +5,64 @@
...
-    # clip values based on alpha
-    if x < -alpha:
-        return -alpha
-    elif x > alpha:
-        return alpha
-
-    levels_lst = list(levels)
-    indices_lst = list(indices)
+    if round(x, 4) in float_to_apot_cache:
+        return float_to_apot_cache[round(x, 4)]
```
Comment (jerryzh168) [inline torch/ao/quantization/experimental/apot_utils.py:18]: wait, doesn't this cache depend on the quantization parameters? i.e. if we change `levels`, 'indicies', 'alpha' this will break right?

File: torch/ao/quantization/experimental/qconfig.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+                                                                                     quant_max=15,
+                                                                                     observer=MinMaxObserver,
+                                                                                     dtype=torch.quint8),
+                                   weight=APoTFakeQuantize.with_args(b=4, k=2, dtype=torch.qint8))
+
+# APoT activation and uniform weight, b=4 k=2
+apot_qconfig_4bit = QConfig(activation=APoTFakeQuantize.with_args(b=4, k=2, dtype=torch.quint8),
+                            weight=APoTFakeQuantize.with_args(b=4, k=2, dtype=torch.qint8))
+
+def get_uniform_qconfig_8bit():
```
Comment (jerryzh168) [inline torch/ao/quantization/experimental/qconfig.py:41]: feels like no need to define these functions, we can just use the variable directly

File: torch/ao/quantization/experimental/apot_utils.py [hunk 6]
```diff
@@ -5,52 +5,64 @@
...
     reduced precision floating point value
     based on quantization levels
 """
-def float_to_reduced_precision(x, levels, indices):
-    levels_lst = list(levels)
-    indices_lst = list(indices)
-
-    min_delta = math.inf
-    best_fp = 0.0
+def float_to_reduced_precision(x, levels):
```
Comment (jerryzh168) [inline torch/ao/quantization/experimental/apot_utils.py:43]: nit: it would clearer if we name this "quant_dequant_apot" I think

File: torch/ao/quantization/experimental/fake_quantize.py [hunk 7]
```diff
@@ -1,34 +1,45 @@
...
+    def __init__(self, observer=APoTObserver, **observer_kwargs):
         super().__init__()
-        self.activation_post_process = APoTObserver(**observer_kwargs)
+        self.activation_post_process = observer(**observer_kwargs)
+        dtype = observer_kwargs.get("dtype", torch.quint8)
+        if hasattr(observer, "p"):
+            # If observer is _PartialWrapper, dtype can be stored in
+            # observer.p.keywords["dtype"]
+            dtype = getattr(getattr(observer, "p", {}), "keywords", {}).get(
+                "dtype", dtype)
```
Comment (jerryzh168) [inline torch/ao/quantization/experimental/fake_quantize.py:21]: I think this can be removed?

File: torch/ao/quantization/experimental/fake_quantize.py [hunk 8]
```diff
@@ -1,34 +1,45 @@
...
             assert (self.alpha is not None
                     and self.gamma is not None
                     and self.quantization_levels is not None
                     and self.level_indices is not None), "Must set qparams for fake quant"
 
-            X = quantize_APoT(X, self.alpha, self.gamma, self.quantization_levels, self.level_indices)
-            X = dequantize_APoT(X)
+            X = quant_dequant_APoT(X, self.alpha, self.gamma, self.quantization_levels, self.level_indices)
 
+        print("finished forward")
```
Comment (jerryzh168) [inline torch/ao/quantization/experimental/fake_quantize.py:44]: remove

File: torch/ao/quantization/experimental/apot_utils.py (hunk 4 above)
Comment (asl3) [reply torch/ao/quantization/experimental/apot_utils.py:18]: oh yes, i was running each instance of quantization parameters independently so it wouldn't be an issue but that would be an issue if they were run together -- will fix!

File: test/quantization/core/experimental/fx_graph_mode_apot.py (hunk 3 above)
Comment (asl3) [reply test/quantization/core/experimental/fx_graph_mode_apot.py:267]: since this is the only instance of eager mode quantization maybe i will leave outside? not sure if a helper function would simplify much

**Comment by dzdang** [comment]
PR title mismatch with summary

File: torch/ao/quantization/fake_quantize.py [hunk 9]
```diff
@@ -328,6 +328,8 @@ def forward(self, X: torch.Tensor) -> torch.Tensor:
 Default fake_quant for activations.
 """
 
+default_symmetric_fake_quant = FakeQuantize.with_args(observer=MovingAverageMinMaxObserver, qscheme=torch.per_tensor_symmetric)
```
Comment (jerryzh168) [inline torch/ao/quantization/fake_quantize.py:331]: would be nice if we specify dtype as well I think, just to be more explicit

**Review by jerryzh168 — APPROVED** [review]
looks good, thanks for addressing the comments
